src/cdk-experimental/listbox/listbox.ts

```diff
@@ -194,360 +158,382 @@ export class CdkOption<T = unknown> implements ListKeyManagerOption, Highlightab
+  /** Emits when the internal value of the listbox changes for any reason. */
+  _internalValueChange = this.selectionModelSubject.pipe(
+    switchMap(selectionModel => selectionModel.changed),
+    takeUntil(this.destroyed),
+  );
+
+  /** Callback called when the listbox has been touched */
+  private _onTouched: () => void = () => {};
+
+  /** Callback called when the listbox value changes */
+  private _onChange: (value: T[]) => void = () => {};
+
+  /** Emits when an option has been clicked. */
+  private _optionClicked = defer(() =>
+    (this.options.changes as Observable<CdkOption<T>[]>).pipe(
+      startWith(this.options),
+      switchMap(options => merge(...options.map(option => option._clicked.pipe(mapTo(option))))),
+    ),
+  );
 
   constructor(
+    /** The host element of the listbox. */
+    protected elementRef: ElementRef,
+    // TODO(mmalerba): Should not depend on combobox
     @Optional() @Inject(CDK_COMBOBOX) private readonly _combobox: CdkCombobox,
+    /** The directionality of the page. */
     @Optional() private readonly _dir?: Directionality,
   ) {}
 
-  ngOnInit() {
-    this._selectionModel = new SelectionModel<CdkOption<T>>(this.multiple);
-  }
-
   ngAfterContentInit() {
     this._initKeyManager();
-    this._initSelectionModel();
     this._combobox?._registerContent(this.id, 'listbox');
-
-    this.optionSelectionChanges.subscribe(event => {
-      this._emitChangeEvent(event.source);
-      this._updateSelectionModel(event.source);
-      this.setActiveOption(event.source);
-      this._updatePanelForSelection(event.source);
-    });
+    this._optionClicked
+      .pipe(
+        filter(option => !option.disabled),
+        takeUntil(this.destroyed),
+      )
+      .subscribe(option => {
+        this.listKeyManager.setActiveItem(option);
+        this.triggerOption(option);
+        this._updatePanelForSelection(option);
+      });
   }
 
   ngOnDestroy() {
-    this._listKeyManager.change.complete();
-    this._destroyed.next();
-    this._destroyed.complete();
+    this.listKeyManager.change.complete();
+    this.destroyed.next();
+    this.destroyed.complete();
   }
 
-  private _initKeyManager() {
-    this._listKeyManager = new ActiveDescendantKeyManager(this._options)
-      .withWrap()
-      .withTypeAhead()
-      .withHomeAndEnd()
-      .withAllowedModifierKeys(['shiftKey']);
-
-    if (this.orientation === 'vertical') {
-      this._listKeyManager.withVerticalOrientation();
-    } else {
-      this._listKeyManager.withHorizontalOrientation(this._dir?.value || 'ltr');
-    }
-
-    this._listKeyManager.change.pipe(takeUntil(this._destroyed)).subscribe(() => {
-      this._updateActiveOption();
-    });
+  /**
+   * Toggle the selected state of the given option.
+   * @param option The option to toggle
+   */
+  toggle(option: CdkOption<T> | T) {
+    this.selectionModel().toggle(option instanceof CdkOption ? option.value : option);
```

Comment:
Shouldn't we require for the options to always extend `CdkOption`? Allowing anything in there may make it harder to introduce features in the future.

Reply:
We should keep directive-on-host-elements in mind for this, where it may not directly extend the class

Reply:
This is meant to allow the user to toggle by option or by value, e.g. if we have:
```html
<ul cdkListbox>
  <li cdkOption="apple" #appleOption="cdkOption">Apple</li>
  <li cdkOption="orange">Orange</li>
</ul>
```

Then the user could do either of these to toggle the apple option:
```ts
@ViewChild(CdkListbox) listbox: CdkListbox<string>;
@ViewChild('appleOption') appleOption: CdkOption<string>;

listbox.toggle(appleOption)
listbox.toggle('apple')
```

I assume in many cases toggle-by-value is simpler. WDYT?

Reply:
I'd be more inclined to have separate methods for this e.g. `toggle(option)` and `toggleValue(value)`